### PR TITLE
Add filter functionality to File System Repository

### LINKED ISSUE #32314: [ProjectAPI] Implement platform compatibility filter in File System
**Description:**
`AbstractPackageRepository` needs to filter out dependencies based on compatability. A util needs to be written that doesnt pull newer dependencies built using newer distributions. 

**Describe your task(s)**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by Ibaqu**
## Purpose
Add a function to filter package version to AbstractPackageRepository. 

The function compares the package version with the version of the pack being used to build the project. Versions greater than the pack version are filtered out. 

Fixes #32314 

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java [hunk 1]
```diff
@@ -172,9 +173,54 @@ public FileSystemRepository(Environment environment, Path cacheDirectory, String
         } catch (IOException e) {
             throw new RuntimeException("Error while accessing Distribution cache: " + e.getMessage());
         }
+
+        List<Path> incompatibleVersions = new ArrayList<>();
+
+        if (!versions.isEmpty()) {
+            for (Path ver : versions) {
+                Path pkgJsonPath = bala.resolve(org.value()).resolve(name.value()).resolve(ver.toString())
+                        .resolve("any").resolve(ProjectConstants.PACKAGE_JSON);
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java:182]: The platform can be `any` or `java11`. Shall we try to make use of https://github.com/ballerina-platform/ballerina-lang/blob/5254c78fc787190bcd9c36135368ecc900803e6a/compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java#L242 method to get the bala path?

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java [hunk 2]
```diff
@@ -172,9 +173,54 @@ public FileSystemRepository(Environment environment, Path cacheDirectory, String
...
+
+                if (Files.exists(pkgJsonPath)) {
+                    String packageVer = BalaFiles.readPkgJson(pkgJsonPath).getBallerinaVersion()
+                            .replace("sl", "");
+                    String packVer = RepoUtils.getBallerinaShortVersion().replace("sl", "");
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java:187]: we can remove `slbeta` when trimming the version string and compare the remaining number

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java (hunk 1 above)
Comment (Ibaqu) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java]: Made use of `getPackagePath` to resolve platform

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java (hunk 2 above)
Comment (Ibaqu) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java]: Removed trimming

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java [hunk 3]
```diff
@@ -172,9 +173,54 @@ public FileSystemRepository(Environment environment, Path cacheDirectory, String
...
+        if (!versions.isEmpty()) {
+            for (Path ver : versions) {
+                Path pkgJsonPath = getPackagePath(org.value(), name.value(), ver.getFileName().toString())
+                        .resolve(ProjectConstants.PACKAGE_JSON);
+
+                if (Files.exists(pkgJsonPath)) {
+                    String packageVer = BalaFiles.readPkgJson(pkgJsonPath).getBallerinaVersion();
+                    String packVer = RepoUtils.getBallerinaShortVersion();
+
+                    if (!packageVer.equals(packVer)) {
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java:188]: We have done this check inside the isCompatible as well? we can remove the from one place.

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java [hunk 4]
```diff
@@ -172,9 +173,54 @@ public FileSystemRepository(Environment environment, Path cacheDirectory, String
...
+            }
+        }
+
+        versions.removeAll(incompatibleVersions);
         return pathToVersions(versions);
     }
 
+    private boolean isCompatible(String packageVer, String packVer) {
+        if (!packageVer.equals(packVer)) {
+            String packageStage = packageVer.substring(0, packageVer.length() - 1);
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java:205]: shall we rename the variables to be more explanatory? e.g. `pkgPlatform` and `distPlatform`

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java [hunk 5]
```diff
@@ -172,9 +173,54 @@ public FileSystemRepository(Environment environment, Path cacheDirectory, String
...
+            String packStage = packVer.substring(0, packVer.length() - 1);
+
+            // If the stages are equal, we need to check the versions
+            if (packageStage.equals(packStage)) {
+                String packageVerValue = packageVer.substring(packageVer.length() - 1);
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java:210]: shall we rename the variables to be more explanatory? e.g. `pkgPlatformVer` and `distPlatformVer`

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java [hunk 6]
```diff
@@ -172,9 +173,54 @@ public FileSystemRepository(Environment environment, Path cacheDirectory, String
...
+                    incompatibleVersions.add(ver);
+                }
+            }
+        }
+
+        versions.removeAll(incompatibleVersions);
         return pathToVersions(versions);
     }
 
+    private boolean isCompatible(String packageVer, String packVer) {
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java:203]: shall we rename the parameters to be more explanatory? e.g. `pkgBalVersion` and `distBalVersion`

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java [hunk 7]
```diff
@@ -172,9 +173,54 @@ public FileSystemRepository(Environment environment, Path cacheDirectory, String
         } catch (IOException e) {
             throw new RuntimeException("Error while accessing Distribution cache: " + e.getMessage());
         }
+
+        List<Path> incompatibleVersions = new ArrayList<>();
+
+        if (!versions.isEmpty()) {
+            for (Path ver : versions) {
+                Path pkgJsonPath = getPackagePath(org.value(), name.value(), ver.getFileName().toString())
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java:181]: `ver.getFileName()` can be `null`. We should wrap it with `Optional.of`

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java (hunk 3 above)
Comment (Ibaqu) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java]: Removed extra checks

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java (hunk 6 above)
Comment (Ibaqu) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java]: Changed param names to pkgBalVersion and distBalVersion

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java (hunk 7 above)
Comment (Ibaqu) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java]: made use of Optional to resolve spotbugs issue

File: compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java [hunk 8]
```diff
@@ -173,9 +174,54 @@ public FileSystemRepository(Environment environment, Path cacheDirectory, String
...
+                }
+            }
+        }
+
+        return incompatibleVersions;
+    }
+
+    private boolean isCompatible(String pkgBalVer, String distBalVer) {
+        if (!pkgBalVer.equals(distBalVer)) {
+            String packageStage = pkgBalVer.substring(0, pkgBalVer.length() - 1);
```
Comment (azinneera) [inline compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java:206]: Shall we rename  this variables to be more explanatory? e.g: `pkgBalVersionPrefix`

Comment (azinneera) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java:206]: We can also update the comments

Comment (Ibaqu) [reply compiler/ballerina-lang/src/main/java/io/ballerina/projects/internal/repositories/FileSystemRepository.java]: Changed the isCompatible() variable names and comments

**Review by azinneera — APPROVED** [review]

**Pull request merged**